### ISSUE TITLE
Too many open log files

### ISSUE THREAD

**Issue by rerooted** [aws/amazon-ssm-agent#323]
When running SSM on an AWS Linux 2 VM (t3a.nano) I suffer from too many open files.  All of the instances are within auto-scaling groups.  There are 12 instances, across two regions and three availability zones. The VMs are running Postfix, Bind or Squid.  Everything on these VMs is standard, only AWS packages are installed.  The load on each of the servers is very low.  Each of VMs running Bind or Postfix connect to the local Squid proxy in the same region.

I have been having this issue for the last few AWS releases, probably going back some months.

```
# lsof | awk '/^ssm-agent/{print $NF}' | sort | uniq -c
[snip]
   6216 /var/log/amazon/ssm/amazon-ssm-agent.log
   2716 /var/log/amazon/ssm/amazon-ssm-agent.log.2
     14 /var/log/amazon/ssm/amazon-ssm-agent.log.3
   5208 /var/log/amazon/ssm/errors.log
[snip]
```

On a freshly booted system I have around 2500 open files.  The number of SSM open log files increases over a few weeks until the error start occurring.

Towards the tail end of the errors.log I have errors about too many open files:

```
... with error - pipe2: too many open files
... too many open files
... with error - open /dev/null: too many open files
```

```
2020-11-11 07:01:02 ERROR [replaceLogger @ ssmlog.go.147] New logger creation failed
2020-11-11 07:01:02 ERROR [replaceLogger @ ssmlog.go.147] New logger creation failed
2020-11-11 07:01:02 ERROR [replaceLogger @ ssmlog.go.147] New logger creation failed
2020-11-11 07:01:02 ERROR [replaceLogger @ ssmlog.go.147] New logger creation failed
```

Package version: amazon-ssm-agent-3.0.161.0-1.amzn2.x86_64

/etc/amazon/ssm/seelog.xml

```
<seelog type="adaptive" mininterval="2000000" maxinterval="100000000" critmsgcount="500" minlevel="info">
    <exceptions>
        <exception filepattern="test*" minlevel="error"/>
    </exceptions>
    <outputs formatid="fmtinfo">
        <console formatid="fmtinfo"/>
        <rollingfile type="size" filename="/var/log/amazon/ssm/amazon-ssm-agent.log" maxsize="30000000" maxrolls="2"/>
        <filter levels="error,critical" formatid="fmterror">
            <rollingfile type="size" filename="/var/log/amazon/ssm/errors.log" maxsize="10000000" maxrolls="2"/>
        </filter>
    </outputs>
    <formats>
        <format id="fmterror" format="%Date %Time %LEVEL [%FuncShort @ %File.%Line] %Msg%n"/>
        <format id="fmtdebug" format="%Date %Time %LEVEL [%FuncShort @ %File.%Line] %Msg%n"/>
        <format id="fmtinfo" format="%Date %Time %LEVEL %Msg%n"/>
    </formats>
</seelog>
```

**Comment by Thor-Bjorgvinsson** [comment]
Thanks for reporting this issue, we have identified an issue where loggers not being closed properly, we are working towards a fix to be included in our next release.

**Comment by Thor-Bjorgvinsson** [comment]
Fix for this has been released in the latest agent release 3.0.502.0. Please allow 2 weeks for it to be available in all regions
